### PR TITLE
[ENH] Update ddtw distance

### PR DESCRIPTION
**Pull request by chrisholder**
#### Reference Issues/PRs
This pr depends on pr #357 

#### What does this implement/fix? Explain your changes.
This pr updates ddtw to the new format for distances. In addition in doing so it makes optimisations for computing the distances between multiple time series. 

#### Did you add any tests for the change?
There are two new test files that extend testing to cost matrix and alignment paths.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc).
- [x] Optionally, I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.
- [x] For new estimators, I've added the estimator to the online documentation.

### CODE REVIEW

File: aeon/distances/_ddtw.py [hunk 1]
```diff
@@ -1,47 +1,344 @@
...
+        _x = average_of_slope(x.reshape((1, x.shape[0])))
+        _y = average_of_slope(y.reshape((1, y.shape[0])))
+        bounding_matrix = create_bounding_matrix(_x.shape[1], _y.shape[1], window)
+        return _dtw_distance(_x, _y, bounding_matrix)
+    if x.ndim == 2 and y.ndim == 2:
+        _x = average_of_slope(x)
+        _y = average_of_slope(y)
+        bounding_matrix = create_bounding_matrix(_x.shape[1], _y.shape[1], window)
+        return _dtw_distance(_x, _y, bounding_matrix)
+    if x.ndim == 3 and y.ndim == 3:
```
Comment (TonyBagnall) [inline aeon/distances/_ddtw.py:79]: I really dont think we need the weird summing up option for 3D, I dont really understand the use case, was there an issue raised for this?

File: aeon/distances/_ddtw.py [hunk 2]
```diff
@@ -1,47 +1,344 @@
...
+    >>> X = np.array([[[10, 22, 399]],[[41, 500, 1316]], [[117, 18, 9]]])
+    >>> y_univariate = np.array([[100, 11, 199],[10, 15, 26], [170, 108, 1119]])
+    >>> ddtw_pairwise_distance(X, y_univariate)
+    array([[ 15129.    ],
+           [322624.    ],
+           [  3220.5625]])
+    """
+    if y is None:
+        # To self
+        if X.ndim == 3:
```
Comment (TonyBagnall) [inline aeon/distances/_ddtw.py:212]: is it possible to adapt this to also work with list of 2D numpy array, so we can also process unequal length collections?

**Review by TonyBagnall — CHANGES_REQUESTED** [review]
looks good, but can it be adapted to work for np-list? We want to be able to model unequal length collections as lists of 2D numpy rather than 3D numpy

File: aeon/distances/_ddtw.py (hunk 1 above)
Comment (chrisholder) [reply aeon/distances/_ddtw.py:79]: There was but if you don't think it's a good idea we can just remove it and replace it with an error

**Review by TonyBagnall — APPROVED** [review]
agreed with @chrisholder to take out the weird 3D thing and refactor for unequal length after all these are in

File: aeon/distances/_ddtw.py [hunk 3]
```diff
@@ -1,47 +1,344 @@
...
+            _x = X.reshape((1, 1, X.shape[0]))
+            _y = y.reshape((y.shape[0], 1, y.shape[1]))
+            return _ddtw_from_multiple_to_multiple_distance(_x, _y, window)
+        else:
+            raise ValueError("x and y must be 2D or 3D arrays")
+
+
+@njit(cache=True, fastmath=True)
+def _ddtw_pairwise_distance(X: np.ndarray, window: float) -> np.ndarray:
+    n_instances = X.shape[0]
```
[the remark this answers is not in the document]
Comment (TonyBagnall) [reply aeon/distances/_ddtw.py:253]: do it after

**Pull request merged**